File: python/01_simple/snake.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Простая консольная змейка с использованием библиотеки curses
Управление: стрелки или WASD, Q - выход
"""

import curses
import random
import time

class SimpleSnakeGame:
    def __init__(self, stdscr):
        self.stdscr = stdscr
        self.setup_game()
        
    def setup_game(self):
        # Настройка экрана
        curses.curs_set(0)  # Скрыть курсор
        self.stdscr.nodelay(1)  # Неблокирующий ввод
        self.stdscr.timeout(100)  # Таймаут в миллисекундах
        
        # Цвета
        curses.start_color()
        curses.init_pair(1, curses.COLOR_GREEN, curses.COLOR_BLACK)  # Змейка
        curses.init_pair(2, curses.COLOR_RED, curses.COLOR_BLACK)    # Еда
        curses.init_pair(3, curses.COLOR_WHITE, curses.COLOR_BLACK)  # Стены
        
        # Размеры поля
        self.height, self.width = self.stdscr.getmaxyx()
        self.height -= 2  # Оставляем место для информации
        self.width -= 2
        
        # Начальные параметры игры
        self.reset_game()
        
    def reset_game(self):
        """Сброс игры к начальному состоянию"""
        # Змейка
        self.snake = [
            [self.height // 2, self.width // 2],
            [self.height // 2, self.width // 2 - 1],
            [self.height // 2, self.width // 2 - 2]
        ]
        
        # Направление движения (вправо)
        self.direction = curses.KEY_RIGHT
        self.score = 0
        self.game_over = False
        
        # Создание еды
        self.create_food()
        
    def create_food(self):
        """Создание еды в случайном месте"""
        while True:
            self.food = [
                random.randint(1, self.height - 2),
                random.randint(1, self.width - 2)
            ]
            if self.food not in self.snake:
                break
                
    def draw_border(self):
        """Отрисовка границ поля"""
        for i in range(self.width + 1):
            self.stdscr.addch(0, i, '═', curses.color_pair(3))
            self.stdscr.addch(self.height, i, '═', curses.color_pair(3))
        
        for i in range(self.height + 1):
            self.stdscr.addch(i, 0, '║', curses.color_pair(3))
            self.stdscr.addch(i, self.width, '║', curses.color_pair(3))
        
        self.stdscr.addch(0, 0, '╔', curses.color_pair(3))
        self.stdscr.addch(0, self.width, '╗', curses.color_pair(3))
        self.stdscr.addch(self.height, 0, '╚', curses.color_pair(3))
        self.stdscr.addch(self.height, self.width, '╝', curses.color_pair(3))
        
    def draw(self):
        """Отрисовка всех элементов игры"""
        self.stdscr.clear()
        self.draw_border()
        
        # Отрисовка змейки
        for i, segment in enumerate(self.snake):
            if i == 0:  # Голова
                self.stdscr.addch(segment[0], segment[1], '●', curses.color_pair(1))
            else:
                self.stdscr.addch(segment[0], segment[1], '○', curses.color_pair(1))
        
        # Отрисовка еды
        self.stdscr.addch(self.food[0], self.food[1], '★', curses.color_pair(2))
        
        # Информация
        info = f" Счёт: {self.score} | Q - выход "
        self.stdscr.addstr(self.height + 1, 0, info, curses.color_pair(3))
        
        if self.game_over:
            game_over_msg = " ИГРА ОКОНЧЕНА! Нажмите R для рестарта "
            self.stdscr.addstr(self.height // 2, (self.width - len(game_over_msg)) // 2, 
                              game_over_msg, curses.A_BOLD)
        
        self.stdscr.refresh()
        
    def update(self):
        """Обновление состояния игры"""
        # Получение ввода
        key = self.stdscr.getch()
        
        if key == ord('q') or key == ord('Q'):
            return False
            
        if self.game_over:
            if key == ord('r') or key == ord('R'):
                self.reset_game()
            return True
            
        # Изменение направления
        if key in [curses.KEY_UP, ord('w'), ord('W')] and self.direction != curses.KEY_DOWN:
            self.direction = curses.KEY_UP
        elif key in [curses.KEY_DOWN, ord('s'), ord('S')] and self.direction != curses.KEY_UP:
            self.direction = curses.KEY_DOWN
        elif key in [curses.KEY_LEFT, ord('a'), ord('A')] and self.direction != curses.KEY_RIGHT:
            self.direction = curses.KEY_LEFT
        elif key in [curses.KEY_RIGHT, ord('d'), ord('D')] and self.direction != curses.KEY_LEFT:
            self.direction = curses.KEY_RIGHT
            
        # Движение змейки
        head = self.snake[0].copy()
        
        if self.direction == curses.KEY_UP:
            head[0] -= 1
        elif self.direction == curses.KEY_DOWN:
            head[0] += 1
        elif self.direction == curses.KEY_LEFT:
            head[1] -= 1
        elif self.direction == curses.KEY_RIGHT:
            head[1] += 1
            
        # Проверка на столкновение со стенами
        if (head[0] <= 0 or head[0] >= self.height or
            head[1] <= 0 or head[1] >= self.width):
            self.game_over = True
            return True
            
        # Проверка на поедание еды
        if head == self.food:
            self.snake.insert(0, head)
            self.score += 10
            self.create_food()
        else:
            self.snake.insert(0, head)
            self.snake.pop()
            
        # Проверка на столкновение с собой
        if self.snake[0] in self.snake[1:]:
            self.game_over = True
            
        return True
        
    def run(self):
        """Главный игровой цикл"""
        running = True
        while running:
            running = self.update()
            self.draw()
            time.sleep(0.05)  # Небольшая задержка

def main(stdscr):
    game = SimpleSnakeGame(stdscr)
    game.run()

if __name__ == "__main__":
    curses.wrapper(main)

File: javascript/01_simple/snake.js
#!/usr/bin/env node

/**
 * Простая консольная змейка для Node.js
 * Управление: WASD, Q - выход
 * 
 * Запуск: node snake.js
 */

const readline = require('readline');

readline.emitKeypressEvents(process.stdin);
process.stdin.setRawMode(true);

class SimpleSnakeGame {
    constructor(width = 20, height = 10) {
        this.width = width;
        this.height = height;
        this.reset();
        this.setupInput();
    }

    reset() {
        // Создание змейки
        this.snake = [
            { x: Math.floor(this.width / 2), y: Math.floor(this.height / 2) }
        ];
        this.direction = 'RIGHT';
        this.nextDirection = 'RIGHT';
        this.score = 0;
        this.gameOver = false;
        this.gameLoop = null;
        
        // Создание еды
        this.createFood();
    }

    createFood() {
        while (true) {
            this.food = {
                x: Math.floor(Math.random() * this.width),
                y: Math.floor(Math.random() * this.height)
            };
            
            // Проверка, что еда не на змейке
            if (!this.snake.some(segment => 
                segment.x === this.food.x && segment.y === this.food.y)) {
                break;
            }
        }
    }

    setupInput() {
        process.stdin.on('keypress', (str, key) => {
            if (key.ctrl && key.name === 'c') {
                process.exit();
            }

            if (key.name === 'q') {
                process.exit();
            }

            if (this.gameOver) {
                if (key.name === 'r') {
                    this.reset();
                }
                return;
            }

            // Управление
            switch (key.name) {
                case 'w': this.nextDirection = 'UP'; break;
                case 's': this.nextDirection = 'DOWN'; break;
                case 'a': this.nextDirection = 'LEFT'; break;
                case 'd': this.nextDirection = 'RIGHT'; break;
            }
        });
    }

    move() {
        // Проверка и применение направления
        const opposite = {
            'UP': 'DOWN',
            'DOWN': 'UP',
            'LEFT': 'RIGHT',
            'RIGHT': 'LEFT'
        };

        if (this.nextDirection !== opposite[this.direction]) {
            this.direction = this.nextDirection;
        }

        // Новая голова
        const head = { ...this.snake[0] };
        
        switch (this.direction) {
            case 'UP': head.y--; break;
            case 'DOWN': head.y++; break;
            case 'LEFT': head.x--; break;
            case 'RIGHT': head.x++; break;
        }

        // Проверка столкновения со стенами
        if (head.x < 0 || head.x >= this.width || 
            head.y < 0 || head.y >= this.height) {
            this.gameOver = true;
            return;
        }

        // Добавление новой головы
        this.snake.unshift(head);

        // Проверка поедания еды
        if (head.x === this.food.x && head.y === this.food.y) {
            this.score += 10;
            this.createFood();
        } else {
            this.snake.pop();
        }

        // Проверка столкновения с собой
        const headCollision = this.snake.slice(1).some(segment => 
            segment.x === head.x && segment.y === head.y
        );

        if (headCollision) {
            this.gameOver = true;
        }
    }

    render() {
        console.clear();
        
        // Верхняя граница
        let output = '┌' + '─'.repeat(this.width) + '┐\n';

        // Игровое поле
        for (let y = 0; y < this.height; y++) {
            output += '│';
            for (let x = 0; x < this.width; x++) {
                if (this.snake[0].x === x && this.snake[0].y === y) {
                    output += '●'; // Голова
                } else if (this.snake.slice(1).some(s => s.x === x && s.y === y)) {
                    output += '○'; // Тело
                } else if (this.food.x === x && this.food.y === y) {
                    output += '★'; // Еда
                } else {
                    output += ' ';
                }
            }
            output += '│\n';
        }

        // Нижняя граница
        output += '└' + '─'.repeat(this.width) + '┘\n';
        
        // Информация
        output += `\n Счёт: ${this.score}\n`;
        output += ` Управление: WASD | Q - выход | R - рестарт\n`;

        if (this.gameOver) {
            output += `\n 🎮 ИГРА ОКОНЧЕНА! Нажмите R для рестарта\n`;
        }

        console.log(output);
    }

    start() {
        console.log('=== КОНСОЛЬНАЯ ЗМЕЙКА ===\n');
        this.gameLoop = setInterval(() => {
            if (!this.gameOver) {
                this.move();
            }
            this.render();
        }, 150);
    }
}

// Запуск игры
const game = new SimpleSnakeGame();
game.start();
